scene/3d/navigation_agent.cpp

```diff
@@ -0,0 +1,374 @@
+	ADD_PROPERTY(PropertyInfo(Variant::REAL, "time_horizon", PROPERTY_HINT_RANGE, "0.01,100,0.01"), "set_time_horizon", "get_time_horizon");
+	ADD_PROPERTY(PropertyInfo(Variant::REAL, "max_speed", PROPERTY_HINT_RANGE, "0.1,10000,0.01"), "set_max_speed", "get_max_speed");
+	ADD_PROPERTY(PropertyInfo(Variant::REAL, "path_max_distance", PROPERTY_HINT_RANGE, "0.01,100,0.1"), "set_path_max_distance", "get_path_max_distance");
+	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "ignore_y"), "set_ignore_y", "get_ignore_y");
+
+	ADD_SIGNAL(MethodInfo("path_changed"));
+	ADD_SIGNAL(MethodInfo("target_reached"));
+	ADD_SIGNAL(MethodInfo("navigation_finished"));
+	ADD_SIGNAL(MethodInfo("velocity_computed", PropertyInfo(Variant::VECTOR3, "safe_velocity")));
+}
+
+void NavigationAgent::_notification(int p_what) {
+	switch (p_what) {
+		case NOTIFICATION_READY: {
+			agent_parent = Object::cast_to<Spatial>(get_parent());
+
+			NavigationServer::get_singleton()->agent_set_callback(agent, this, "_avoidance_done");
+
+			// Search the navigation node and set it
+			{
+				Navigation *nav = nullptr;
+				Node *p = get_parent();
+				while (p != nullptr) {
+					nav = Object::cast_to<Navigation>(p);
+					if (nav != nullptr) {
+						p = nullptr;
+					}
+					else {
+						p = p->get_parent();
+					}
+				}
+
+				set_navigation(nav);
+			}
+
+			set_physics_process_internal(true);
+		} break;
+		case NOTIFICATION_EXIT_TREE: {
+			agent_parent = nullptr;
+			set_navigation(nullptr);
+			set_physics_process_internal(false);
+			// Want to call ready again when the node enters the tree again. We're not using enter_tree notification because
+			// the navigation map may not be ready at that time. This fixes issues with taking the agent out of the scene tree.
+			request_ready();
+		} break;
+		case NOTIFICATION_INTERNAL_PHYSICS_PROCESS: {
+			if (agent_parent) {
+				NavigationServer::get_singleton()->agent_set_position(agent, agent_parent->get_global_transform().origin);
+				_check_distance_to_target();
+			}
+		} break;
+	}
+}
+
+NavigationAgent::NavigationAgent() :
+		agent_parent(nullptr),
+		navigation(nullptr),
+		agent(RID()),
+		target_desired_distance(1.0),
+		navigation_height_offset(0.0),
+		path_max_distance(3.0),
+		velocity_submitted(false),
+		target_reached(false),
+		navigation_finished(true) {
+	agent = NavigationServer::get_singleton()->agent_create();
+	set_neighbor_dist(50.0);
+	set_max_neighbors(10);
+	set_time_horizon(5.0);
+	set_radius(1.0);
+	set_max_speed(10.0);
+	set_ignore_y(true);
+}
+
+NavigationAgent::~NavigationAgent() {
+	NavigationServer::get_singleton()->free(agent);
+	agent = RID(); // Pointless
+}
+
+void NavigationAgent::set_navigation(Navigation *p_nav) {
+	if (navigation == p_nav)
+		return; // Pointless
+
+	navigation = p_nav;
+	NavigationServer::get_singleton()->agent_set_map(agent, navigation == nullptr ? RID() : navigation->get_rid());
```

Comment:
Nitpick, but this is shorter :P
```suggestion
	NavigationServer::get_singleton()->agent_set_map(agent, navigation ? navigation->get_rid() : RID());
```

Reply:
Shorter? Yes, but a lot of people have issues with ternary operators so I think it being explicit here is better for readability personally. If you'd like for it to be shorter, sure.

Reply:
Well to me it's as explicit and more readable (at least that's a very common pattern in the engine code), but it doesn't matter much. Can also stay as is.

Reply:
Well, if it is common in the engine code, so be it. I don't really care either way. Updated.